### PR TITLE
apf: manage request queue wait with context

### LINKED ISSUE #119799: apf queue wait time for a request exceeds the hard-coded 15s threshold
### What happened?

The user receives a timeout `504` for a certain request:
```
I0726 22:18:10.778212      11 httplog.go:129] "HTTP" verb="LIST" URI="/api/v1/namespaces/<ns>/pods?
pretty=true&allowWatchBookmarks=false&labelSelector=<selector>&resourceVersion=862900435&watch=false" 
latency="1m0.037950467s" userAgent="Kubernetes Java Client/16.0.0-SNAPSHOT" 
audit-ID="caa77d18-9f66-4d8e-a29d-146a2efbb22c" srcIP="10.100.150.226:45302" 
apf_pl="workload-low" apf_fs="service-accounts" resp=504 statusStack=
```

This is the stack trace from httplog:
```
goroutine 32409485 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).recordStatus(0xc0ff829340, 0x0?)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:314 +0x90
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.(*respLogger).WriteHeader(0xc0ff829340, 0x1?)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:296 +0x25
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*baseTimeoutWriter).timeout(0xc084b08690, 0xc0da50ebe0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:255 +0xd8
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP(0xc00e854978, {0x5713ec8, 0xc08c402740}, 0xdf8475800?)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/timeout.go:146 +0x2e3
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.withRequestDeadline.func1({0x5713ec8, 0xc08c402740}, 0xc050286c00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/request_deadline.go:100 +0x362
net/http.HandlerFunc.ServeHTTP(0x1?, {0x5713ec8?, 0xc08c402740?}, 0x8?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.withWaitGroup.func1({0x5713ec8?, 0xc08c402740}, 0xc050286c00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/waitgroup.go:77 +0x766
net/http.HandlerFunc.ServeHTTP(0xc00383daf0?, {0x5713ec8?, 0xc08c402740?}, 0xc0abaab680?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*goaway).ServeHTTP(0xc01017d100, {0x5713ec8, 0xc08c402740}, 0xc050286c00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/goaway.go:72 +0x146
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuditAnnotations.func1({0x5713ec8, 0xc08c402740}, 0xc050286a00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/audit_annotations.go:36 +0x125
net/http.HandlerFunc.ServeHTTP(0x5714e78?, {0x5713ec8?, 0xc08c402740?}, 0x56cf548?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithWarningRecorder.func1({0x5713ec8?, 0xc08c402740}, 0xc050286600)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/warning.go:35 +0x18d
net/http.HandlerFunc.ServeHTTP(0x481b860?, {0x5713ec8?, 0xc08c402740?}, 0xd?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithCacheControl.func1({0x5713ec8, 0xc08c402740}, 0xc090260f68?)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/cachecontrol.go:31 +0x126
net/http.HandlerFunc.ServeHTTP(0x5715f18?, {0x5713ec8?, 0xc08c402740?}, 0x56cf548?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog.withLogging.func1({0x5705a98, 0xc090260f68}, 0xc050286400)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/httplog/httplog.go:128 +0x339
net/http.HandlerFunc.ServeHTTP(0x5714e78?, {0x5705a98?, 0xc090260f68?}, 0x56cf548?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1({0x5705a98, 0xc090260f68}, 0xc050286000)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x1ed
net/http.HandlerFunc.ServeHTTP(0x5714e78?, {0x5705a98?, 0xc090260f68?}, 0x10?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.withRequestReceivedTimestampWithClock.func1({0x5705a98, 0xc090260f68}, 0xc052439f00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/request_received_time.go:38 +0x167
net/http.HandlerFunc.ServeHTTP(0xc0cf0282c0?, {0x5705a98?, 0xc090260f68?}, 0x0?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithMuxAndDiscoveryComplete.func1({0x5705a98, 0xc090260f68}, 0xc052439f00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/mux_discovery_complete.go:52 +0x191
net/http.HandlerFunc.ServeHTTP(0xc00e580000?, {0x5705a98?, 0xc090260f68?}, 0x8?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.withPanicRecovery.func1({0x5705a98?, 0xc090260f68?}, 0xc0cf0282c0?)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters/wrap.go:74 +0xb1
net/http.HandlerFunc.ServeHTTP(0x481b860?, {0x5705a98?, 0xc090260f68?}, 0x8?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.withAuditID.func1({0x5705a98, 0xc090260f68}, 0xc052439e00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/with_auditid.go:66 +0x2f1
net/http.HandlerFunc.ServeHTTP(0xc0abaabc98?, {0x5705a98?, 0xc090260f68?}, 0x3e59880?)
	/usr/local/go/src/net/http/server.go:2109 +0x2f
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.(*APIServerHandler).ServeHTTP(0xc0f1ab6d3f?, {0x5705a98?, 0xc090260f68?}, 0x5?)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/handler.go:189 +0x2b
net/http.serverHandler.ServeHTTP({0x699880?}, {0x5705a98, 0xc090260f68}, 0xc052439e00)
	/usr/local/go/src/net/http/server.go:2947 +0x30c
net/http.initALPNRequest.ServeHTTP({{0x5714e78?, 0xc0f3196e40?}, 0xc057590e00?, {0xc0091c8e10?}}, {0x5705a98, 0xc090260f68}, 0xc052439e00)
	/usr/local/go/src/net/http/server.go:3556 +0x245
k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*serverConn).runHandler(0x815d928?, 0x0?, 0x0?, 0xc0abaabfb0?)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2311 +0x83
created by k8s.io/kubernetes/vendor/golang.org/x/net/http2.(*serverConn).processHeaders
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/golang.org/x/net/http2/server.go:2025 +0x64a
" addedInfo="
logging error output: \"{\\\"metadata\\\":{},\\\"status\\\":\\\"Failure\\\",\\\"message\\\":\\\
"Timeout: request did not complete within the allotted timeout\\\",\\\
"reason\\\":\\\"Timeout\\\",\\\"details\\\":{},\\\"code\\\":504}\
\"
"
```

The timeout filter aborted the request and sent a `504` to the caller:
https://github.com/kubernetes/kubernetes/blob/953be8927218ec8067e1af2641e540238ffd7576/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L146

The goroutine that the timeout handler spawned to execute the request (rest of the handler chain) was in flight for more than `60s`, and it is recording a `429` response code, implying that apf rejected the request.
```
{
  "kind": "Event",
  "apiVersion": "audit.k8s.io/v1",
  "level": "Request",
  "auditID": "caa77d18-9f66-4d8e-a29d-146a2efbb22c",
  "stage": "ResponseComplete",
  "requestURI": "/api/v1/namespaces/<ns>/pods?pretty=true&allowWatchBookmarks=false&labelSelector=<selector>&resourceVersion=862900435&watch=false",
  "verb": "list",
  "user": {
    "username": "system:serviceaccount:<sa>",
    "uid": "5ab073d0-5c0e-4985-bca3-20038108d401",
    "groups": [
      "system:serviceaccounts",
      "system:serviceaccounts:<sa>",
      "system:authenticated"
    ],
    "extra": {
      "authentication.kubernetes.io/pod-name": [
        "<pod-name>"
      ],
      "authentication.kubernetes.io/pod-uid": [
        "11916349-428e-4d9f-9167-899fcacbf6d5"
      ]
    }
  },
  "sourceIPs": [
    "10.100.150.226"
  ],
  "userAgent": "Kubernetes Java Client/16.0.0-SNAPSHOT",
  "objectRef": {
    "resource": "pods",
    "namespace": "<ns>",
    "apiVersion": "v1"
  },
  "responseStatus": {
    "metadata": {},
    "code": 429
  },
  "requestReceivedTimestamp": "2023-07-26T22:17:10.740109Z",
  "stageTimestamp": "2023-07-26T22:18:10.904195Z"
}
```

This means that the request was waiting in the queue for `60s` before it was removed from the queue and rejected. We are already aware of the fact that the removal check happens upon arrival of a new request for the priority level. But we only check the queue that has been selected for the newly arrived request, there could be other queues in the queueset that could have old requests that are candidate for rejection.

When a new request arrives, we assign it to a queue first and then remove old requests from the selected queue. This means, the logic for selecting a queue for the new request takes into account the work in seat-seconds of these "should be rejected now" requests waiting in the set of queues being considered. 

So a queue with more work in seat-seconds is less likely to be selected and this phenomenon may contribute to a request waiting for more than `15s`
Also there is the extreme case of new request not arriving for the given priority level.

https://github.com/kubernetes/kubernetes/blob/1620473a9a01dd6bbef3398c0acb2e581d0a13c3/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L568-L576

https://github.com/kubernetes/kubernetes/blob/1620473a9a01dd6bbef3398c0acb2e581d0a13c3/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L619-L620

Other observations:
- Maximum observed time from audit logs was 61.9922s out of the thousands of examples
- The given priority level (`workload-low`) was maxed out, requests that failed to be dispatched were rejected with reasons cancelled or time-out 

How we can avoid this issue:
A: when a new request arrives:
- we scan all the queues of the queueset and reject the requests that have been waiting for more than `15s` 
- we select a queue for the new request
(we depend on new request arrival for the queueset, as we do today, and iterating trough the requests waiting in the queues for the queueset is costly to the O(n), n is the number of requests waiting)

B: we can use the context of the request, if it has a deadline (true for all non long-running requests - it's the `timeout` specified in the request URI, otherwise `60s`
- we derive a new context from the request context, and pass the new context to the promise that will be used by req.wait(). the new context has a deadline of 1/4 of the request timeout (`15s` at most)
This approach does not depend on arrival of new request - https://github.com/kubernetes/kubernetes/pull/105849 is a work in progress for this approach.

### What did you expect to happen?

Ideally, queue wait time of a request should not exceed its deadline or does not depend on arrival of any future request.

### How can we reproduce it (as minimally and precisely as possible)?

Hard to reproduce manually, but audit entry and httplog from production shows it's happening

### Kubernetes version

v1.23.17

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
manage request queue wait with deadline bound context

#### Which issue(s) this PR fixes:
Fixes #119799

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by dims** [comment]
please reopen if you still need this WIP PR. Trying to declutter the open PR(s)

**Comment by wojtek-t** [comment]
/assign

Will try to get it asap, but have a bunch of things to catch up.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 1]
```diff
@@ -227,6 +247,7 @@ func (qsc *queueSetCompleter) Complete(dCfg fq.DispatchingConfig) fq.QueueSet {
 			requestsExecutingSet:     sets.New[*request](),
 		}
 		qs.promiseFactory = qsc.factory.promiseFactoryFactory(qs)
+		qs.queueWaitContextFactory = qsc.factory.queueWaitContextFactory
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:250]: Why not do this in the constructor just above?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go [hunk 2]
```diff
@@ -556,44 +574,41 @@ func (qs *queueSet) getVirtualTimeRatioLocked() float64 {
...
 // returns the enqueud request on a successful enqueue
 // returns nil in the case that there is no available concurrency or
 // the queuelengthlimit has been reached
-func (qs *queueSet) timeoutOldRequestsAndRejectOrEnqueueLocked(ctx context.Context, workEstimate *fqrequest.WorkEstimate, hashValue uint64, flowDistinguisher, fsName string, descr1, descr2 interface{}, queueNoteFn fq.QueueNoteFn) *request {
+func (qs *queueSet) rejectOrEnqueueLocked(ctx context.Context, workEstimate *fqrequest.WorkEstimate, hashValue uint64, flowDistinguisher, fsName string, descr1, descr2 interface{}, queueNoteFn fq.QueueNoteFn) *request {
 	// Start with the shuffle sharding, to pick a queue.
 	queueIdx := qs.shuffleShardLocked(hashValue, descr1, descr2)
 	queue := qs.queues[queueIdx]
-	// The next step is the logic to reject requests that have been waiting too long
-	qs.removeTimedOutRequestsFromQueueToBoundLocked(queue, fsName)
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go:574]: Was this the only reason for the following call to `qs.boundNextDispatchLocked`?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go [hunk 3]
```diff
@@ -1037,7 +1026,7 @@ func TestDifferentFlowsWithoutQueuing(t *testing.T) {
 	now := time.Now()
 
 	clk, counter := testeventclock.NewFake(now, 0, nil)
-	qsf := newTestableQueueSetFactory(clk, countingPromiseFactoryFactory(counter))
+	qsf := newTestableQueueSetFactory(clk, countingPromiseFactoryFactory(counter), test.NewFakeQueueWait())
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1029]: This test was using RequestWaitLimit=0

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queuewait_test.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+			queueWaitCtxGot, cancelGot := qw.GetQueueWaitContext(parent, test.defaultQueueWaitTime)
+			if cancelGot == nil {
+				t.Errorf("Expected a non nil context.CancelFunc")
+				return
+			}
+			defer cancelGot()
+
+			switch {
+			case test.queueWaitCtxExpected:
+				if _, ok := queueWaitCtxGot.Deadline(); !ok {
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queuewait_test.go:91]: Wouldn't it simplify the regular code and the test if this test examined the deadline and parent equality instead of hijacking internal details of `qw`?  The simpler way would be a black-box test that gets you the same code coverage.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/promise/counting.go [hunk 5]
```diff
@@ -41,40 +43,52 @@ type countingPromise struct {
...
+	duration time.Duration
+}
+
+func NewQueueWaitTimeWithContext(parent context.Context, d time.Duration) context.Context {
+	ctx := context.WithValue(parent, testKey, &testData{duration: d})
+	return ctx
+}
+
+func GetQueueWaitTimeFromContext(ctx context.Context) (time.Duration, bool) {
+	if td, ok := ctx.Value(testKey).(*testData); ok {
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/promise/counting.go:60]: Don't you want to do that cast only if `ok` is true?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go [hunk 6]
```diff
@@ -17,12 +17,13 @@ limitations under the License.
 package promise
 
 import (
+	"context"
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go:20]: Can we move the first commit to a separate PR - it's non-controversial at all.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go:1029]: The question is whether it's something we really want to test it - do we use it in production?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queuewait.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		// 'defaultQueueWaitTime' which is 15s today. (This is in keeping with
+		// the current behavior)
+		queueWaitTime := deadline.Sub(qw.clock.Now()) / 4
+		if queueWaitTime <= 0 {
+			return parent, func() {}
+		}
+		return qw.newCtxFn(parent, queueWaitTime)
+	}
+
+	return qw.newCtxFn(parent, defaultQueueWaitTime)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queuewait.go:67]: If defaultQueueWaitTime would be smaller than 1/4th of deadline, should we choose that then? Or do we ignore it completely?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queuewait_test.go (hunk 4 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queuewait_test.go:91]: +1

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/fake_queuewait.go [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+	"time"
+
+	testingpromise "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/promise"
+)
+
+func NewFakeQueueWait() fakeQueueWait {
+	return fakeQueueWait{}
+}
+
+type fakeQueueWait struct{}
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/fake_queuewait.go:30]: It would be good to declare the interface that this type is intended to implement.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/promise/counting_test.go [hunk 9]
```diff
@@ -36,109 +36,114 @@ func TestMain(m *testing.M) {
...
-	cval := &oldTime
-	doneCh := make(chan struct{})
+	doneCtx := NewQueueWaitTimeWithContext(context.Background(), time.Second)
 	now := time.Now()
 	clock, counter := testeventclock.NewFake(now, 0, nil)
 	var lock sync.Mutex
-	wr := NewCountingWriteOnce(counter, &lock, nil, doneCh, cval)
+	wr := NewCountingWriteOnce(doneCtx, clock, counter, &lock, nil, "canceled")
+
+	beforeCanceled := now.Add(time.Second).Add(-time.Millisecond)
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/promise/counting_test.go:45]: You could do `now.Add(time.Second - time.Millisecond)`

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queuewait.go [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+// newQueueWaitContextFactory returns a new instance of queueWaitContextFactory
+func newQueueWaitContextFactory(clock eventclock.Interface) *queueWaitFactory {
+	return &queueWaitFactory{
+		clock:    clock,
+		newCtxFn: context.WithTimeout,
+	}
+}
+
+type queueWaitFactory struct {
```
Comment (MikeSpreitzer) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queuewait.go:34]: It would be good to declare (in a way that the compiler checks) the interface that this type is intended to implement.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go (hunk 6 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go:20]: Actually I see no reason for this change.  A context has a Done channel.

**Comment by MikeSpreitzer** [comment]
Before this change we had these two among the possible values of type `requestDecision`:

```go
	// Reject this one due to APF queuing considerations
	decisionReject

	// This one's context timed out / was canceled
	decisionCancel
```

If I understand correctly, the thrust of this PR is to make sure that there is always a queuing time limit and it is no more generous than the request's incoming time limit.  In other words, the queuing time limit may expire but there is no other timeout that can happen.  To my mind this means that `decisionReject` is kept and this PR should remove `decisionCacnel`.  But the current revision of the PR does the opposite.  It matters to metrics, which are exposed to users.

**Comment by MikeSpreitzer** [comment]
I suspect there is a simpler approach to solving the problem that this PR addresses, just using a Context with a possibly nearer deadline in the existing time-limiting logic in the QueueSet.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go (hunk 6 above)
Comment (tkashem) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/promise/promise.go:20]: the context is not required for the production code, i used the context in the current test code to propagate queue wait time (there may be alternatives, but that would require changes in queueset.go as well)

https://github.com/kubernetes/kubernetes/pull/120162 (if we want it)

**Comment by tkashem** [comment]
> I suspect there is a simpler approach to solving the problem that this PR addresses, just using a Context with a possibly nearer deadline in the existing time-limiting logic in the QueueSet.

most of the changes in this PR are around fixing the existing tests to cope with queue wait time - the change in queueset.go seems reasonably small, imo. If there is a better alternate, I am happy to incorporate it.
I already moved one commit to a separate PR, maybe i can look into moving the fixes to the tests in separate PRs., do you think that will help?

**Comment by wojtek-t** [comment]
> most of the changes in this PR are around fixing the existing tests to cope with queue wait time - the change in queueset.go seems reasonably small, imo. If there is a better alternate, I am happy to incorporate it.
I already moved one commit to a separate PR, maybe i can look into moving the fixes to the tests in separate PRs., do you think that will help?

I personally like the fact that requests are independent and are taking care of removing themselves from the queue (TBH I like it better then the current approach, even if we ignore the existing bug we're trying to fix).
But let's discuss it during upcoming APF meeting maybe.

Regarding splitting - are there test changes that we can do independently (without code changes)? If so, let's separate them...

**Comment by tkashem** [comment]
> Regarding splitting - are there test changes that we can do independently (without code changes)? If so, let's separate them...

Yes, currently the tests do not exercise queue wait time (except for one, if i am not mistaken). So I can open a new PR with the test fixes i have made in this PR.  This will significantly reduce the size of this PR, i believe

**Comment by tkashem** [comment]
converting this to a draft for now, in favor of https://github.com/kubernetes/kubernetes/pull/120222 which has the same approach but circumvents the test issue by moving the context creation logic outside of queueset.

120222 seems easier to backport, so please take a look when you have a moment.
